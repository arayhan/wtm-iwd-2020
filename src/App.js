import React, { Component } from 'react'
import firebase from 'firebase'
import { Fade } from "react-reveal"
import { withRouter } from "react-router-dom"
import LandingPage from './Pages/LandingPage/LandingPage'
import Loading from './Components/Atoms/Loading'
import MemberArea from './Pages/MemberArea/MemberArea'
import { db } from "./Services/Firebase"
import './Services/Firebase'

class App extends Component {
	constructor(props) {
		super(props)
		this.state = {
			isSignedIn: false,
			isLoading: true,
			isCompleteProfile: false,
			user: {},
		}
	}

	componentDidMount = async () => {
		await firebase.auth().onAuthStateChanged(async (user) => {
			if (user) {
				try {
					await db.ref('users').orderByChild('uid').equalTo(user.uid)
						.on('value', snapshot => {
							let userData = []
							let isCompleteProfile
							if (snapshot.exists()) {
								isCompleteProfile = true
								snapshot.forEach((snap) => {
									userData.push(snap.val())
								})
							}
							else {
								isCompleteProfile = false
								userData = [{
									uid: user.uid,
									displayName: user.displayName,
									email: user.email,
									photoURL: user.photoURL,
								}]
							}
							this.props.history.push('/')
							this.setState({
								user: userData[0],
								isCompleteProfile: isCompleteProfile,
								isSignedIn: user,
								isLoading: false,
							})
						})
				} catch (error) {
					this.setState({
						readError: error.message, loadingChats: false,
						isSignedIn: user,
						isLoading: false,
					})
				}
			} else {
				this.setState({
					isSignedIn: user,
					isLoading: false,
				})
			}
		})
	};

	setUserProfile = (user) => {
		this.setState({
			user: user,
			isCompleteProfile: true
		})
	};

	render() {
		const { isSignedIn, isLoading, isCompleteProfile, user } = this.state
		return (
			<div>
				<Loading isLoading={isLoading} />

				{!isSignedIn &&
					<Fade when={!isLoading}>
						<LandingPage />
					</Fade>
				}

				{isSignedIn &&
					<Fade when={!isLoading}>
						<MemberArea
							isCompleteProfile={isCompleteProfile}
							user={user}
							setUserProfile={() => this.setUserProfile()}
						/>
					</Fade>
				}
			</div>
		)
	}
}

export default withRouter(App)
