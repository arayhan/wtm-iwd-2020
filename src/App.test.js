import React from 'react';
import App from './App';
import ShallowRenderer from 'react-test-renderer/shallow';
import firebase from 'firebase';
import {shallow} from "enzyme";

let component;

jest.mock('firebase', () => ({
  initializeApp: jest.fn(),
  analytics: jest.fn(),
  signInOptions: {},
  auth: jest.fn(),
  onAuthStateChanged: jest.fn(),
}));

jest.mock('./Config/Congfig', () => ({
  uiConfig: {},
}));

jest.mock('./Config/Recognition', () => ({
  recognition: {},
}));

beforeEach(() => {
  firebase.auth.mockReturnValue({
    onAuthStateChanged: jest.fn()
  });
  component = shallow(<App/>);
});

describe('render', () => {
  it('should render loading section when isSignedIn state is null', () => {
    component.setState({isSignedIn: null});

    const loadingSection = component.find('#loading-section');

    expect(loadingSection).toHaveLength(1);
  });

  it('should render login section when isSignedIn state is false', () => {
    component.setState({isSignedIn: false});

    const loginSection = component.find('#login-section');

    expect(loginSection).toHaveLength(1);
  });

  it('should render dashboard section when isSignedIn state is true', () => {
    component.setState({isSignedIn: true});

    const dashboardSection = component.find('#dashboard-section');

    expect(dashboardSection).toHaveLength(1);
  });
});


describe('snapshot', () => {
  it('should same like snapshot', () => {
    const shallowRenderer = new ShallowRenderer();
    const result = shallowRenderer.render(<App/>);
    expect(result).toMatchSnapshot();
  });
});

describe('componentDidMount', () => {
  it('should set state isSignedIn to true when user auth state has been changed to true', () => {
    firebase.auth().onAuthStateChanged.mockImplementationOnce(() =>{
      component.setState({
        isSignedIn: true,
      })
    });

    component.instance().componentDidMount();

    expect(component.state().isSignedIn).toEqual(true);
  });
});
