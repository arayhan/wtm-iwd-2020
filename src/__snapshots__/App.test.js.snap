// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should same like snapshot 1`] = `
<div
  className="App"
>
  <NavbarTop />
  <div
    className="center-screen"
    id="loading-section"
  >
    <Fade>
      <hr />
      <h1>
        Loading ...
      </h1>
      <hr />
    </Fade>
  </div>
</div>
`;
